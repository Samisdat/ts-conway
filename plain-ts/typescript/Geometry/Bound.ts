export class Bound {

    private min: number;

    private max: number;

    constructor(min: number, max: number) {

        if (max < min) {
            throw new Error('max can not be lower then min');
        }

        this.min = min;
        this.max = max;
    }

    public isWithin(test: number): boolean {
        return (test >= this.min && test <= this.max);
    }

    public isBelow(test: number): boolean {
        return (test < this.min);
    }

    public isAbove(test: number): boolean {
        return (test > this.max);
    }

    public confine(value: number): number {

        if (value < this.min) {
            return this.min;
        }

        if (value > this.max) {
            return this.max;
        }

        return value;

    }

    public getMax(): number{
        return this.max;
    }

    public getMin(): number{
        return this.min;
    }

    public expand(value: number): void{

        if(true === this.isWithin(value)){
            return;
        }

        if(true === this.isBelow(value)){
            this.min = value
        }

        if(true === this.isAbove(value)){
            this.max = value
        }

    }

}
