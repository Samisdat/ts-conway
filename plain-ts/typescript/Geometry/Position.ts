export class Position {

    public readonly  x: number;
    public readonly  y: number;

    constructor(x: number, y: number) {

        if (-0 === x) {
            x = 0;
        }

        if (-0 === y) {
            y = 0;
        }

        this.x = x;
        this.y = y;
    }

    public move(position: Position): Position {
        return new Position(
            this.x + position.x,
            this.y + position.y
        );
    }

    public inverse(): Position {
        return new Position(
            -1 * this.x,
            -1 * this.y
        );
    }

    public compare(position: Position): boolean {

        return (position.x === this.x && position.y === this.y);

    }

    public getNeighbours(): Position[] {

        const neighbours: Position[] = [];

        // neighbours on bottom
        neighbours.push(
            new Position(this.x - 1, this.y - 1)
        );
        neighbours.push(
            new Position(this.x, this.y - 1)
        );
        neighbours.push(
            new Position(this.x + 1, this.y - 1)
        );

        // neighbours on the left and on the right
        neighbours.push(
            new Position(this.x - 1, this.y)
        );
        neighbours.push(
            new Position(this.x + 1, this.y)
        );

        // neighbours on top
        neighbours.push(
            new Position(this.x - 1, this.y + 1)
        );
        neighbours.push(
            new Position(this.x, this.y + 1)
        );
        neighbours.push(
            new Position(this.x + 1, this.y + 1)
        );

        return neighbours;
    }

    public clone(): Position {
        return new Position(
            this.x,
            this.y
        );
    }

    public toString(): string {

        let str = '';

        str += (0 <= this.x) ? '+' : '';
        str += this.x;

        str += '/';

        str += (0 <= this.y) ? '+' : '';
        str += this.y;

        return str;

    }
}
