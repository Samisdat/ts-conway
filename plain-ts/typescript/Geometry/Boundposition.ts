import {Position} from '@Conway/Geometry/Position';
import {Bound} from '@Conway/Geometry/Bound';

export class Boundposition {

    private x: Bound;
    private y: Bound;

    constructor(topLeft: Position, bottomRight: Position) {

        if (bottomRight.x < topLeft.x) {
            throw new Error('bottomRight must be right from topLeft');
        }

        if (bottomRight.y < topLeft.y) {
            throw new Error('bottomRight must be below from topLeft');
        }

        this.x = new Bound(topLeft.x, bottomRight.x);
        this.y = new Bound(topLeft.y, bottomRight.y);

    }

    public isWithin(test: Position): boolean {
        return (this.x.isWithin(test.x) && this.y.isWithin(test.y));
    }

    public confine(value: Position): Position {

        return new Position(
            this.x.confine(value.x),
            this.y.confine(value.y)
        );

    }

    public bottomLeft(): Position{

        return new Position(this.x.getMin(), this.y.getMin());

    }

    public topRight(): Position{

        return new Position(this.x.getMax(), this.y.getMax());

    }

    public getWidth(): number{

        return this.x.getMax() - this.x.getMin() + 1;

    }

    public getHeight(): number{

        return this.y.getMax() - this.y.getMin() + 1;

    }

    public expand(position: Position): void{

        if(true === this.isWithin(position)){
            return;
        }

        this.x.expand(position.x);
        this.y.expand(position.y);

    }

}
