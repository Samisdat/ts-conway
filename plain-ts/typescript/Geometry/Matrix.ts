import {Position} from '@Conway/Geometry/Position';
import {Boundposition} from '@Conway/Geometry/Boundposition';

interface MatrixInterface {
    [index: string]: Position;
}


export class Matrix {

    private matrix: MatrixInterface = {};

    private boundposition: Boundposition;

    public has(position: Position): boolean {

        return (undefined !== this.matrix[position.toString()]);

    }

    public add(position: Position): void {

        if (true === this.has(position)) {
            return;
        }

        this.matrix[position.toString()] = position;


        if(undefined === this.boundposition){
            this.boundposition = new Boundposition(position, position);
        }
        else{
            this.boundposition.expand(position);
        }

    }

    public remove(position: Position): void {

        if (false === this.has(position)) {
            return;
        }

        delete this.matrix[position.toString()];

    }

    public getAll(): Position[] {

        const positions: Position[] = [];

        for (const index in this.matrix) {

            positions.push(this.matrix[index]);

        }

        return positions;

    }

    public getBound(): Boundposition {

        return this.boundposition;

    }

    public getWidth(): number {

        return this.getBound().getWidth();

    }

    public getHeight(): number {

        return this.getBound().getHeight();

    }

}
