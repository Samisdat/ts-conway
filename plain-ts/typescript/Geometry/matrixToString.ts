import {Matrix} from '@Conway/Geometry/Matrix';
import {Position} from '@Conway/Geometry/Position';

interface NumberedScale {
    sign: string[];
    number: string[];
};

const getSigns = (numberedScale: NumberedScale, i: number) =>{
    if(0 === i % 5){

        numberedScale.number.push(Math.abs(i).toString());

        if(0 > i){
            numberedScale.sign.push('-');
        }
        else if(0 < i){
            numberedScale.sign.push('+');
        }
        else{
            numberedScale.sign.push(' ');
        }

    }
    else{
        numberedScale.number.push(' ');
        numberedScale.sign.push(' ');
    }

};

const getScaleHorizontal = (matrix: Matrix, matrixArray: string[][]): string[][] => {

    const left = matrix.getBound().bottomLeft().x;
    const right  = matrix.getBound().topRight().x;

    matrixArray = matrixArray.reverse();

    const newRow: string[] = []
    for(let i = 0; i < matrixArray[0].length; i += 1){
        newRow.push('')
    }

    matrixArray.push(newRow);

    const scale: NumberedScale = {
        sign: [],
        number: [],
    };

    for(let i = left; i <= right; i += 1){

        getSigns(scale, i);
    }

    matrixArray.push(scale.number);
    matrixArray.push(scale.sign);

    matrixArray = matrixArray.reverse();
    return matrixArray;


};

const getScaleVertical = (matrix: Matrix, matrixArray: string[][]): string[][] => {

    const top  = matrix.getBound().topRight().y;
    const bottom = matrix.getBound().bottomLeft().y;

    const scale: NumberedScale = {
        sign: [' ', ' ', ' '],
        number: [' ', ' ', ' '],
    };

    for(let i = top; i >= bottom; i -= 1){

        getSigns(scale, i);

    }

    for(let i = 0; i < matrixArray.length; i += 1){

        matrixArray[i] = matrixArray[i].reverse();

        matrixArray[i].push(' ')
        matrixArray[i].push(' ')

        matrixArray[i].push(scale.number[i])
        matrixArray[i].push(scale.sign[i])

        matrixArray[i] = matrixArray[i].reverse();

    }

    return matrixArray;

};

const addScale = (matrix: Matrix, matrixString: string[][]): string[][] => {

    matrixString = getScaleHorizontal(matrix, matrixString);
    matrixString = getScaleVertical(matrix, matrixString);
    return matrixString;
};

export const matrixToString = (matrix: Matrix, scale = true): string => {

    if(0 === matrix.getAll().length){
        return 'Matrix is empty';
    }

    const width = matrix.getWidth();
    const height = matrix.getHeight();

    const topLeft = new Position(
        matrix.getBound().bottomLeft().x,
        matrix.getBound().topRight().y
    );

    let matrixOfStrings: string[][] = [];

    let pointer = topLeft.clone();

    for(let y = 0; y < height; y += 1){

        const line = [];

        for(let x = 0; x < width; x += 1){

            if(true === matrix.has(pointer)){
                line.push('O');
            }
            else{
                line.push('.');
            }

            pointer = pointer.move(new Position(1,0));

        }

        matrixOfStrings.push(line);

        pointer = new Position(
            topLeft.x,
            pointer.y - 1
        );

    }

    if(true === scale){
        matrixOfStrings = addScale(matrix, matrixOfStrings);
    }

    return matrixOfStrings.map((row)=>{return row.join('')}).join('\n');;

}
