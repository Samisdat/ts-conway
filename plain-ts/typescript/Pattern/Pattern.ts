import {Position} from '@Conway/Geometry/Position';
import {readPatternFromPlainFile} from '@Conway/Pattern/readPatternFromPlainFile';
import {Matrix} from '@Conway/Geometry/Matrix';
import {patternToString} from '@Conway/Pattern/patternToString';

export class Pattern {

    private name: string;

    private matrix: Matrix;

    constructor(name: string, matrix: Matrix) {

        this.name = name;

        this.matrix = matrix;

    }

    public static fromString(patternString: string): Pattern {
        return readPatternFromPlainFile(patternString);
    }

    public getName(): string {
        return this.name;
    }

    public getMatrix(): Matrix {

        return this.matrix;
    }

    public getWidth(): number {
        return this.matrix.getWidth();
    }

    public getHeight(): number {
        return this.matrix.getHeight();
    }

    public mirrorVertical(): void {

        const mirror = new Matrix();

        for (const position of this.matrix.getAll()) {

            mirror.add(
                new Position(
                    position.x,
                    -1 * position.y
                )
            );

        }

        this.matrix = mirror;

    }

    public mirrorHorizontal(): void {

        const mirror = new Matrix();

        for (const position of this.matrix.getAll()) {

            mirror.add(
                new Position(
                    -1 * position.x,
                    position.y
                )
            );

        }

        this.matrix = mirror;

    }

    public rotate(): void {

        const matrix = this.toArray();

        const rotated: number[][] = [];

        for (let x = 0; x < matrix[0].length; x += 1) {

            const row: number[] = [];

            for (let y = matrix.length - 1; y >= 0 ; y -= 1) {

                row.push(matrix[y][x]);

            }

            rotated.push(row);

        }
        const rotatedMatrix = new Matrix();

        for (let y = 0; y < rotated.length; y += 1) {

            for (let x = 0; x < rotated[y].length; x += 1) {

                if (0 === rotated[y][x]) {
                    continue;
                }

                rotatedMatrix.add(new Position(x, y));

            }

        }

        this.matrix = rotatedMatrix;

    }

    public toArray(): number[][] {

        const moved = new Matrix();

        const moveBy = this.getMatrix().getBound().bottomLeft().inverse();

        for (const position of this.matrix.getAll()) {

            moved.add(
                position.move(moveBy)
            );

        }

        const matrixAsArray: number[][] = [];

        for (let y = 0; y < this.getHeight() + 1; y += 1) {

            const row: number[] = [];

            for (let x = 0; x < this.getWidth() + 1; x += 1) {

                const cell: number = (true === moved.has(new Position(x, y))) ? 1 : 0;

                row.push(cell);

            }

            matrixAsArray.push(row);
        }

        return matrixAsArray;

    }

    public toString(): string {

        return patternToString(this);

    }

}
