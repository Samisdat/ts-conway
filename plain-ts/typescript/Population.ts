import {Position} from '@Conway/Geometry/Position';
import {Matrix} from '@Conway/Geometry/Matrix';

export class Population {

    private readonly generationDuration: number;

    private matrix: Matrix;

    constructor(
        matrix: Matrix,
        generationDuration: number
    ) {

        this.matrix = matrix;

        this.generationDuration = generationDuration;

    }

    public startAging(): void {

        setInterval(
            () => {
                this.elapse();
            },
            this.generationDuration
        );

    }

    private isLiving(position: Position): boolean {

        return this.matrix.has(position);

    }

    private countLivingNeighbours(position: Position): number {

        let livingNeighbours = 0;

        const neighbours: Position[] = position.getNeighbours();

        for (const neighbour of neighbours) {

            const isLiving = this.isLiving(neighbour);

            if (true === isLiving) {
                livingNeighbours += 1;
            }

        }

        return livingNeighbours;

    }

    private applyRules(): void {

        const nextGeneration = new Matrix();

        /**
         * there are four rules
         * 1) Any live cell with fewer than two live neighbours dies, as if caused by under-population.
         * 2) Any live cell with two or three live neighbours lives on to the next generation.
         * 3) Any live cell with more than three live neighbours dies, as if by over-population.
         * 4) Any dead cell with exactly three live neighbours becomes a live cell, as if by reproduction.
         *
         *  All rules dealing with living cells can be handled by kill all cells that don't have two or three neighbours
         *  That's the first loop
         *
         *  To check if any death cells will be birth, we count the living neighbour of all cells neighbours
         *  That's the second loop
         */

        for (const position of this.matrix.getAll()) {

            const livingNeighbours = this.countLivingNeighbours(position);

            if (2 === livingNeighbours || 3 === livingNeighbours) {

                nextGeneration.add(position);

            }

        }

        for (const position of this.matrix.getAll()) {

            const neighbours = position.getNeighbours();

            for (const neighbour of neighbours) {

                if (true === this.matrix.has(neighbour)) {
                    continue;
                }

                const livingNeighbours = this.countLivingNeighbours(neighbour);

                if (3 === livingNeighbours) {

                    nextGeneration.add(neighbour);

                }

            }

        }

        this.matrix = nextGeneration;

    }

    public elapse(): void {

        this.applyRules();

    }

    public getAllCells(): Position[] {

        return this.matrix.getAll();

    }

    get(): Position[] {

        return this.matrix.getAll();

    }

    public getMatrix(): Matrix {
        return this.matrix;
    }

}
