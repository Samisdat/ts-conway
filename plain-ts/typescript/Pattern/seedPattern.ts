
import {Pattern} from '@Conway/Pattern/Pattern';
import {Position} from '@Conway/Geometry/Position';
import {Matrix} from '@Conway/Geometry/Matrix';

export const seedPattern = (matrix: Matrix, pattern: Pattern, moveBy = new Position(0, 0)): void => {

    const positions = pattern.getMatrix().getAll();

    const offSet = new Position(
        -1 * Math.floor(pattern.getMatrix().getWidth() / 2),
        -1 * Math.floor(pattern.getMatrix().getHeight() / 2)
    ).move(moveBy);

    for (const position of positions) {

        matrix.add(position.move(moveBy));

    }

}
