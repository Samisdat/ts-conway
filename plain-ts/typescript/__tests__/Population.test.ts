import serializer from '../../jest-serialize-conway';
import {Population} from '../Population';
import {Position} from '@Conway/Geometry/Position';
import {blinker} from '@Conway/Pattern/Store/blinker.cells';
import {Pattern} from '@Conway/Pattern/Pattern';
import {seedPattern} from '@Conway/Pattern/seedPattern';
import {Matrix} from '@Conway/Geometry/Matrix';

describe('Habitat', () => {

    let matrix: Matrix;

    beforeEach(function () {

        expect.addSnapshotSerializer(serializer);

        matrix = new Matrix();

    });

    test('can be created', () => {

        const population = new Population(
            matrix,
            1000
        );

        expect(population).toBeInstanceOf(Population);

    });

    test('a not existing cell is not alive', () => {

        matrix.add(
            new Position(0, 1)
        );

        const population = new Population(
            matrix,
            1000
        );

        expect(population.getAllCells().length).toBe(1);

        population.elapse();

        expect(population.getAllCells().length).toBe(0);

    });


    test('snapshot', () => {

        seedPattern(
            matrix,
            Pattern.fromString(blinker)
        );

        const population = new Population(
            matrix,
            1000
        );

        expect(population.getMatrix()).toMatchSnapshot();

        population.elapse();

        expect(population.getMatrix()).toMatchSnapshot();

        population.elapse();

        expect(population.getMatrix()).toMatchSnapshot();
    });

    test('elapse with one cell', () => {

        matrix.add(
            new Position(0, 1)
        );

        const population = new Population(
            matrix,
            1000
        );

        expect(population.get()).toStrictEqual([new Position(0, 1)]);

    });

    test('elapse with a blinker', () => {

        matrix.add(
            new Position(0, 0)
        );
        matrix.add(
            new Position(0, 1)
        );
        matrix.add(
            new Position(0, 2)
        );

        const population = new Population(
            matrix,
            1000
        );

        expect(population.get()).toStrictEqual([
            new Position(0, 0),
            new Position(0, 1),
            new Position(0, 2)
        ]);

        population.elapse();

        expect(population.get()).toStrictEqual([
            new Position(0, 1),
            new Position(-1, 1),
            new Position(1, 1)
        ]);

        // check garbage collection
        expect(population.getAllCells().length).toBe(3);


        population.elapse();

        expect(population.get()).toStrictEqual([
            new Position(0, 1),
            new Position(0, 0),
            new Position(0, 2)
        ]);

        expect(population.getAllCells().length).toBe(3);

    });

});

jest.useFakeTimers();

describe('Habitat aging with interval', function() {

    let population: Population;

    beforeEach(() => {

        const matrix = new Matrix();

        seedPattern(
            matrix,
            Pattern.fromString(blinker)
        );

        population = new Population(
            matrix,
            50
        );

    });

    test('should increase position', function() {

        expect(population.getMatrix()).toMatchSnapshot();

        population.startAging();

        jest.advanceTimersByTime(51);

        expect(population.getMatrix()).toMatchSnapshot();

    });

});
