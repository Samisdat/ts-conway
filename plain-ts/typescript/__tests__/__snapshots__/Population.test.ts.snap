// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Habitat aging with interval should increase position 1`] = `
XXX
OOO
OOO
`;

exports[`Habitat aging with interval should increase position 2`] = `
OX
OX
OX
OO
`;

exports[`Habitat snapshot 1`] = `
XXX
OOO
OOO
`;

exports[`Habitat snapshot 2`] = `
OX
OX
OX
OO
`;

exports[`Habitat snapshot 3`] = `
XXX
OOO
OOO
`;
